packages/block-editor/src/components/inserter/block-patterns-tab.js

```diff
@@ -43,91 +46,115 @@ function BlockPatternsSearchResults( { filterValue, onInsert } ) {
 	}
 }
 
-function BlockPatternsPerCategories( { onInsert } ) {
-	const [ patterns, categories, onClick ] = usePatternsState( onInsert );
+function BlockPatternsCategory( {
+	selectedCategory,
+	onClickCategory,
+	onInsert,
+} ) {
+	const [ allPatterns, allCategories, onClick ] = usePatternsState(
+		onInsert,
+		selectedCategory
+	);
+
+	const [ currentCategoryPatterns, setCategoryPatterns ] = useState( [] );
```

Comment:
Hey, in terms of code organisation, there's a few things I'm not sure I understand:

Why do we need `selectedCategory, onClickCategory,` props outside of this component? 
can we just  replace this line above with

```
const [ selectedCategoryName, setSelectedCategoryName ] = useState( [] );
```

and just compute the patterns to render using a useMemo call: something like

```
const patterns = useMemo(() => 
  allPatterns.filter(pattern => selectedCategoryName === 'uncategorized'
     ? getPatternIndex( pattern ) === Infinity
     : pattern.categories && pattern.categories.includes( selectedCategoryName ),
   [allPatterns, selectedCategoryName] 
);
```

I'd probably also do the exact same thing in `BlockPatternsSearchResults` but would also include the "searchFilter" in the filtering.

Reply:
The reason for saving the current patterns to state is because useAsyncList requires the object reference for a given selection to remain constant between renders, but useMemo may achieve the same, will have another look at it.

Reply:
> Why do we need `selectedCategory, onClickCategory,` props outside of this component?

`selectedCategory` being passed into `usePatternsState` was just me failing to tidy up previous implementation, gone now.  

`onClickCategory` is being used to push up the category changes to the parent component so the current category is preserved between tab toggles, which may not be the best approach ... is there a common pattern for cases where you want to maintain state changes in a child tab between tab toggles, but state is minor enough to warrant a full blown redux flow?

Reply:
> onClickCategory is being used to push up the category changes to the parent component so the current category is preserved between tab toggles. Is it acceptable to lift state up like this?

Og right! seems like a good pattern here if it's an expectation for us. In Gutenberg, we tend to use the store for these kind of things but this works too.

Reply:
@youknowriad  - I have switched to using useMemo as you suggested to compute the patterns per render for both the category list and search, rather than storing the computed pattern set in state. 

I have left it with lifting the selected pattern to the parent for now to avoid complicating the PR further, but we could move it to using the store for this in a future PR if we thought it was worth doing this.